src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpInterop.cs

```diff
@@ -0,0 +1,313 @@
+
+            [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "RegisterWasiPollableHandle")]
+            static extern Task CallRegisterWasiPollableHandle(Thread t, int handle, CancellationToken cancellationToken);
+        }
+
+        public static Method ConvertMethod(HttpMethod requestMethod)
+        {
+            Method method;
+            switch (requestMethod.Method)
+            {
+                case "":
+                case "GET":
+                    method = Method.get();
+                    break;
+                case "HEAD":
+                    method = Method.head();
+                    break;
+                case "POST":
+                    method = Method.post();
+                    break;
+                case "PUT":
+                    method = Method.put();
+                    break;
+                case "DELETE":
+                    method = Method.delete();
+                    break;
+                case "CONNECT":
+                    method = Method.connect();
+                    break;
+                case "OPTIONS":
+                    method = Method.options();
+                    break;
+                case "TRACE":
+                    method = Method.trace();
+                    break;
+                case "PATCH":
+                    method = Method.patch();
+                    break;
+                default:
+                    method = Method.other(requestMethod.Method);
+                    break;
+            }
+            return method;
+        }
+
+        public static Scheme ConvertScheme(Uri uri)
+        {
+            Scheme scheme;
+            switch (uri.Scheme)
+            {
+                case "":
+                case "http":
+                    scheme = Scheme.http();
+                    break;
+                case "https":
+                    scheme = Scheme.https();
+                    break;
+                default:
+                    scheme = Scheme.other(uri.Scheme);
+                    break;
+            }
+            return scheme;
+        }
+
+        public static string ConvertAuthority(Uri uri)
+        {
+            // `wasi:http/outgoing-handler` requires a non-empty authority,
+            // so we set one here:
+            if (string.IsNullOrEmpty(uri.Authority))
+            {
+                if (uri.Scheme == "https")
+                {
+                    return ":443";
+                }
+                else
+                {
+                    return ":80";
+                }
+            }
+            else
+            {
+                return uri.Authority;
+            }
+        }
+
+        public static Fields ConvertRequestHeaders(HttpRequestMessage request)
+        {
+            var headers = new List<(string, byte[])>();
+            foreach (var pair in request.Headers)
+            {
+                foreach (var value in pair.Value)
```

Comment:
are we iterating here through a "pair", that is, through 2-items long collection?

Reply:
No, each header could have many values